src/Raven.Server/Rachis/Leader.cs

```diff
@@ -929,19 +929,26 @@ public void Dispose()
 
                         foreach (var entry in _entries)
                         {
-                            if (entry.Value.OnNotify != null)
+                            if (entry.Key <= _lastCommit)
                             {
-                                try
+                                if (entry.Value.OnNotify != null)
                                 {
-                                    entry.Value.OnNotify(entry.Value.TaskCompletionSource);
+                                    try
+                                    {
+                                        entry.Value.OnNotify(entry.Value.TaskCompletionSource);
+                                    }
+                                    catch (Exception e)
+                                    {
+                                        entry.Value.TaskCompletionSource.TrySetException(e);
+                                        continue;
+                                    }
                                 }
-                                catch (Exception e)
+                                else
                                 {
-                                    entry.Value.TaskCompletionSource.TrySetException(e);
-                                    continue;
+                                    entry.Value.TaskCompletionSource.TrySetResult((entry.Value.CommandIndex, entry.Value.Result));
```

Comment:
Shouldn't we also cancel the `TaskCompletionSource` here if this is `> _lastCommit` ?

Reply:
No, because 'TrySetResult' turns the task into 'RanToCompletion' state (final state).